sql/core/src/main/scala/org/apache/spark/sql/internal/CatalogImpl.scala

```diff
@@ -524,14 +527,17 @@ class CatalogImpl(sparkSession: SparkSession) extends Catalog {
     // If this table is cached as an InMemoryRelation, drop the original
     // cached version and make the new version cached lazily.
     val cache = sparkSession.sharedState.cacheManager.lookupCachedData(table)
+
+    // uncache the logical plan.
+    // note this is a no-op for the table itself if it's not cached, but will invalidate all
+    // caches referencing this table.
+    sparkSession.sharedState.cacheManager.uncacheQuery(table, cascade = true)
```

Comment:
Add it to the migration guide?

Reply:
hmm which migration guide should I add this to? note Spark already invalidate cache when the target table is cached, and this just extend it to the case when the table itself is not cached.

Reply:
I think he meant https://github.com/apache/spark/blob/master/docs/sql-migration-guide.md to update. I tend to agree that it's safe to mention it unless it's obviously a bug fix.

Reply:
Thanks for the pointer @HyukjinKwon . I'll update there.

Reply:
Created #30256

Reply:
Thanks @sunchao